File: sdk/typescript/src/namespaces/orchestration.ts
/**
 * Orchestration Namespace API
 *
 * Provides methods for unified multi-agent deliberation orchestration:
 * - Async and sync deliberation endpoints
 * - Knowledge context integration
 * - Output channel routing
 * - Template-based workflows
 *
 * Endpoints:
 *   POST /api/v1/orchestration/deliberate      - Async deliberation
 *   POST /api/v1/orchestration/deliberate/sync - Sync deliberation
 *   GET  /api/v1/orchestration/status/:id      - Get status
 *   GET  /api/v1/orchestration/templates       - List templates
 */

/**
 * Team selection strategy options.
 */
export type TeamStrategy = 'specified' | 'best_for_domain' | 'diverse' | 'fast' | 'random';

/**
 * Output format options.
 */
export type OutputFormat = 'standard' | 'decision_receipt' | 'summary' | 'github_review' | 'slack_message';

/**
 * Deliberation status values.
 */
export type DeliberationStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

/**
 * Knowledge source configuration.
 */
export interface KnowledgeSource {
  type: string;
  id: string;
  filters?: Record<string, unknown>;
}

/**
 * Output channel configuration.
 */
export interface OutputChannel {
  type: string;
  id: string;
  options?: Record<string, unknown>;
}

/**
 * Deliberation template.
 */
export interface DeliberationTemplate {
  name: string;
  description: string;
  default_agents?: string[];
  default_rounds?: number;
  output_format?: OutputFormat;
  knowledge_sources?: string[];
}

/**
 * Deliberation result.
 */
export interface DeliberationResult {
  request_id: string;
  status: DeliberationStatus;
  question: string;
  consensus?: string;
  dissents?: string[];
  agents_used: string[];
  rounds_completed: number;
  output_format: OutputFormat;
  output_channels_notified?: string[];
  created_at: string;
  completed_at?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Options for deliberation requests.
 */
export interface DeliberateOptions {
  /** The question or decision to deliberate on */
  question: string;
  /** Sources for context (e.g., "slack:C123", "confluence:12345") */
  knowledgeSources?: Array<string | KnowledgeSource>;
  /** Workspace IDs to include in context */
  workspaces?: string[];
  /** Strategy for agent team selection */
  teamStrategy?: TeamStrategy;
  /** Explicit list of agents to use */
  agents?: string[];
  /** Channels to route results to (e.g., "slack:C789") */
  outputChannels?: Array<string | OutputChannel>;
  /** Format for the output */
  outputFormat?: OutputFormat;
  /** Whether consensus is required (default true) */
  requireConsensus?: boolean;
  /** Priority level (low, normal, high, critical) */
  priority?: string;
  /** Maximum deliberation rounds (default 3) */
  maxRounds?: number;
  /** Timeout for the deliberation in seconds (default 300) */
  timeoutSeconds?: number;
  /** Template name to use */
  template?: string;
  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Client interface for making HTTP requests.
 */
interface OrchestrationClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Orchestration API namespace.
 *
 * Provides methods for unified multi-agent deliberation:
 * - Submit deliberations (async or sync)
 * - Check deliberation status
 * - List available templates
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Async deliberation
 * const result = await client.orchestration.deliberate({
 *   question: 'Should we migrate to Kubernetes?',
 *   knowledgeSources: ['confluence:12345', 'slack:C123456'],
 *   outputChannels: ['slack:C789'],
 * });
 *
 * // Check status
 * const status = await client.orchestration.getStatus(result.request_id);
 * console.log(`Status: ${status.status}`);
 *
 * // Sync deliberation (waits for completion)
 * const syncResult = await client.orchestration.deliberateSync({
 *   question: 'Which testing framework should we use?',
 *   agents: ['claude', 'gpt-4'],
 *   maxRounds: 3,
 * });
 * console.log(`Consensus: ${syncResult.consensus}`);
 *
 * // List available templates
 * const templates = await client.orchestration.listTemplates();
 * ```
 */
export class OrchestrationAPI {
  constructor(private client: OrchestrationClientInterface) {}

  // =========================================================================
  // Deliberation
  // =========================================================================

  /**
   * Submit an async deliberation request.
   *
   * Returns immediately with a request_id that can be used to check status.
   *
   * @param options - Deliberation options
   * @returns Request ID and initial status
   */
  async deliberate(options: DeliberateOptions): Promise<{
    request_id: string;
    status: DeliberationStatus;
    message: string;
  }> {
    const data = this.buildDeliberatePayload(options);

    return this.client.request('POST', '/api/v1/orchestration/deliberate', {
      json: data,
    });
  }

  /**
   * Submit a synchronous deliberation request.
   *
   * Blocks until deliberation completes and returns the full result.
   *
   * @param options - Deliberation options
   * @returns Full deliberation result
   */
  async deliberateSync(options: DeliberateOptions): Promise<DeliberationResult> {
    const data = this.buildDeliberatePayload(options);

    return this.client.request('POST', '/api/v1/orchestration/deliberate/sync', {
      json: data,
    });
  }

  /**
   * Build the request payload for deliberation.
   */
  private buildDeliberatePayload(options: DeliberateOptions): Record<string, unknown> {
    const data: Record<string, unknown> = { question: options.question };

    if (options.knowledgeSources) data.knowledge_sources = options.knowledgeSources;
    if (options.workspaces) data.workspaces = options.workspaces;
    if (options.teamStrategy && options.teamStrategy !== 'best_for_domain') {
      data.team_strategy = options.teamStrategy;
    }
    if (options.agents) data.agents = options.agents;
    if (options.outputChannels) data.output_channels = options.outputChannels;
    if (options.outputFormat && options.outputFormat !== 'standard') {
      data.output_format = options.outputFormat;
    }
    if (options.requireConsensus === false) {
      data.require_consensus = options.requireConsensus;
    }
    if (options.priority && options.priority !== 'normal') {
      data.priority = options.priority;
    }
    if (options.maxRounds !== undefined && options.maxRounds !== 3) {
      data.max_rounds = options.maxRounds;
    }
    if (options.timeoutSeconds !== undefined && options.timeoutSeconds !== 300) {
      data.timeout_seconds = options.timeoutSeconds;
    }
    if (options.template) data.template = options.template;
    if (options.metadata) data.metadata = options.metadata;

    return data;
  }

  // =========================================================================
  // Status
  // =========================================================================

  /**
   * Get the status of a deliberation request.
   *
   * @param requestId - The deliberation request ID
   * @returns Deliberation status and result (if completed)
   */
  async getStatus(requestId: string): Promise<DeliberationResult> {
    return this.client.request('GET', `/api/v1/orchestration/status/${requestId}`);
  }

  // =========================================================================
  // Templates
  // =========================================================================

  /**
   * List available deliberation templates.
   *
   * @returns List of templates with count
   */
  async listTemplates(): Promise<{
    templates: DeliberationTemplate[];
    count: number;
  }> {
    return this.client.request('GET', '/api/v1/orchestration/templates');
  }
}

File: sdk/python/aragora/namespaces/orchestration.py
"""
Orchestration Namespace API.

Provides methods for unified multi-agent deliberation orchestration:
- Async and sync deliberation endpoints
- Knowledge context integration
- Output channel routing
- Template-based workflows

Endpoints:
    POST /api/v1/orchestration/deliberate      - Async deliberation
    POST /api/v1/orchestration/deliberate/sync - Sync deliberation
    GET  /api/v1/orchestration/status/:id      - Get status
    GET  /api/v1/orchestration/templates       - List templates
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

TeamStrategy = Literal["specified", "best_for_domain", "diverse", "fast", "random"]
OutputFormat = Literal["standard", "decision_receipt", "summary", "github_review", "slack_message"]


class OrchestrationAPI:
    """
    Synchronous Orchestration API.

    Provides methods for unified multi-agent deliberation:
    - Submit deliberations (async or sync)
    - Check deliberation status
    - List available templates

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> # Async deliberation
        >>> result = client.orchestration.deliberate(
        ...     question="Should we migrate to Kubernetes?",
        ...     knowledge_sources=["confluence:12345", "slack:C123456"],
        ...     output_channels=["slack:C789"],
        ... )
        >>> # Check status
        >>> status = client.orchestration.get_status(result["request_id"])
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # =========================================================================
    # Deliberation
    # =========================================================================

    def deliberate(
        self,
        question: str,
        knowledge_sources: list[str | dict[str, Any]] | None = None,
        workspaces: list[str] | None = None,
        team_strategy: TeamStrategy = "best_for_domain",
        agents: list[str] | None = None,
        output_channels: list[str | dict[str, Any]] | None = None,
        output_format: OutputFormat = "standard",
        require_consensus: bool = True,
        priority: str = "normal",
        max_rounds: int = 3,
        timeout_seconds: float = 300.0,
        template: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Submit an async deliberation request.

        Returns immediately with a request_id that can be used to check status.

        Args:
            question: The question or decision to deliberate on.
            knowledge_sources: Sources for context (e.g., "slack:C123", "confluence:12345").
            workspaces: Workspace IDs to include in context.
            team_strategy: Strategy for agent team selection.
            agents: Explicit list of agents to use.
            output_channels: Channels to route results to (e.g., "slack:C789").
            output_format: Format for the output.
            require_consensus: Whether consensus is required.
            priority: Priority level (low, normal, high, critical).
            max_rounds: Maximum deliberation rounds.
            timeout_seconds: Timeout for the deliberation.
            template: Template name to use.
            metadata: Additional metadata.

        Returns:
            Dict with request_id and status.
        """
        data: dict[str, Any] = {"question": question}

        if knowledge_sources:
            data["knowledge_sources"] = knowledge_sources
        if workspaces:
            data["workspaces"] = workspaces
        if team_strategy != "best_for_domain":
            data["team_strategy"] = team_strategy
        if agents:
            data["agents"] = agents
        if output_channels:
            data["output_channels"] = output_channels
        if output_format != "standard":
            data["output_format"] = output_format
        if not require_consensus:
            data["require_consensus"] = require_consensus
        if priority != "normal":
            data["priority"] = priority
        if max_rounds != 3:
            data["max_rounds"] = max_rounds
        if timeout_seconds != 300.0:
            data["timeout_seconds"] = timeout_seconds
        if template:
            data["template"] = template
        if metadata:
            data["metadata"] = metadata

        return self._client.request("POST", "/api/v1/orchestration/deliberate", json=data)

    def deliberate_sync(
        self,
        question: str,
        knowledge_sources: list[str | dict[str, Any]] | None = None,
        workspaces: list[str] | None = None,
        team_strategy: TeamStrategy = "best_for_domain",
        agents: list[str] | None = None,
        output_channels: list[str | dict[str, Any]] | None = None,
        output_format: OutputFormat = "standard",
        require_consensus: bool = True,
        priority: str = "normal",
        max_rounds: int = 3,
        timeout_seconds: float = 300.0,
        template: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Submit a synchronous deliberation request.

        Blocks until deliberation completes and returns the full result.

        Args:
            question: The question or decision to deliberate on.
            knowledge_sources: Sources for context.
            workspaces: Workspace IDs to include.
            team_strategy: Strategy for agent team selection.
            agents: Explicit list of agents to use.
            output_channels: Channels to route results to.
            output_format: Format for the output.
            require_consensus: Whether consensus is required.
            priority: Priority level.
            max_rounds: Maximum deliberation rounds.
            timeout_seconds: Timeout for the deliberation.
            template: Template name to use.
            metadata: Additional metadata.

        Returns:
            Full deliberation result.
        """
        data: dict[str, Any] = {"question": question}

        if knowledge_sources:
            data["knowledge_sources"] = knowledge_sources
        if workspaces:
            data["workspaces"] = workspaces
        if team_strategy != "best_for_domain":
            data["team_strategy"] = team_strategy
        if agents:
            data["agents"] = agents
        if output_channels:
            data["output_channels"] = output_channels
        if output_format != "standard":
            data["output_format"] = output_format
        if not require_consensus:
            data["require_consensus"] = require_consensus
        if priority != "normal":
            data["priority"] = priority
        if max_rounds != 3:
            data["max_rounds"] = max_rounds
        if timeout_seconds != 300.0:
            data["timeout_seconds"] = timeout_seconds
        if template:
            data["template"] = template
        if metadata:
            data["metadata"] = metadata

        return self._client.request("POST", "/api/v1/orchestration/deliberate/sync", json=data)

    # =========================================================================
    # Status
    # =========================================================================

    def get_status(self, request_id: str) -> dict[str, Any]:
        """
        Get the status of a deliberation request.

        Args:
            request_id: The deliberation request ID.

        Returns:
            Dict with status and result (if completed).
        """
        return self._client.request("GET", f"/api/v1/orchestration/status/{request_id}")

    # =========================================================================
    # Templates
    # =========================================================================

    def list_templates(self) -> dict[str, Any]:
        """
        List available deliberation templates.

        Returns:
            Dict with templates list and count.
        """
        return self._client.request("GET", "/api/v1/orchestration/templates")


class AsyncOrchestrationAPI:
    """Asynchronous Orchestration API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # =========================================================================
    # Deliberation
    # =========================================================================

    async def deliberate(
        self,
        question: str,
        knowledge_sources: list[str | dict[str, Any]] | None = None,
        workspaces: list[str] | None = None,
        team_strategy: TeamStrategy = "best_for_domain",
        agents: list[str] | None = None,
        output_channels: list[str | dict[str, Any]] | None = None,
        output_format: OutputFormat = "standard",
        require_consensus: bool = True,
        priority: str = "normal",
        max_rounds: int = 3,
        timeout_seconds: float = 300.0,
        template: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit an async deliberation request."""
        data: dict[str, Any] = {"question": question}

        if knowledge_sources:
            data["knowledge_sources"] = knowledge_sources
        if workspaces:
            data["workspaces"] = workspaces
        if team_strategy != "best_for_domain":
            data["team_strategy"] = team_strategy
        if agents:
            data["agents"] = agents
        if output_channels:
            data["output_channels"] = output_channels
        if output_format != "standard":
            data["output_format"] = output_format
        if not require_consensus:
            data["require_consensus"] = require_consensus
        if priority != "normal":
            data["priority"] = priority
        if max_rounds != 3:
            data["max_rounds"] = max_rounds
        if timeout_seconds != 300.0:
            data["timeout_seconds"] = timeout_seconds
        if template:
            data["template"] = template
        if metadata:
            data["metadata"] = metadata

        return await self._client.request("POST", "/api/v1/orchestration/deliberate", json=data)

    async def deliberate_sync(
        self,
        question: str,
        knowledge_sources: list[str | dict[str, Any]] | None = None,
        workspaces: list[str] | None = None,
        team_strategy: TeamStrategy = "best_for_domain",
        agents: list[str] | None = None,
        output_channels: list[str | dict[str, Any]] | None = None,
        output_format: OutputFormat = "standard",
        require_consensus: bool = True,
        priority: str = "normal",
        max_rounds: int = 3,
        timeout_seconds: float = 300.0,
        template: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a synchronous deliberation request."""
        data: dict[str, Any] = {"question": question}

        if knowledge_sources:
            data["knowledge_sources"] = knowledge_sources
        if workspaces:
            data["workspaces"] = workspaces
        if team_strategy != "best_for_domain":
            data["team_strategy"] = team_strategy
        if agents:
            data["agents"] = agents
        if output_channels:
            data["output_channels"] = output_channels
        if output_format != "standard":
            data["output_format"] = output_format
        if not require_consensus:
            data["require_consensus"] = require_consensus
        if priority != "normal":
            data["priority"] = priority
        if max_rounds != 3:
            data["max_rounds"] = max_rounds
        if timeout_seconds != 300.0:
            data["timeout_seconds"] = timeout_seconds
        if template:
            data["template"] = template
        if metadata:
            data["metadata"] = metadata

        return await self._client.request(
            "POST", "/api/v1/orchestration/deliberate/sync", json=data
        )

    # =========================================================================
    # Status
    # =========================================================================

    async def get_status(self, request_id: str) -> dict[str, Any]:
        """Get the status of a deliberation request."""
        return await self._client.request("GET", f"/api/v1/orchestration/status/{request_id}")

    # =========================================================================
    # Templates
    # =========================================================================

    async def list_templates(self) -> dict[str, Any]:
        """List available deliberation templates."""
        return await self._client.request("GET", "/api/v1/orchestration/templates")
